_posts/2019-09-01-exposing-the-commerce-engine.md

```diff
@@ -0,0 +1,34 @@
+
+![Expose using commerce connect and custom endpoint](/assets/images/expose-commerce-engine/commerceconnect.jpg))
+
+
+## Expose engine through API gateway
+Alternatively you could choose to directly use the commerce engine API from JSS. This would safe you from having to create a custom API. Also, although Sitecore Commerce Connect adds flexibility and extensibility, we have noticed in past projects that this also adds complexity because of the mapping from the commerce connect concepts to those of the commerce engine. This is why we really wanted to experiment with using the engine directly from JSS. The challenge here is that the engine API does not seem to have been build to support this scenario. For example, how do we make sure anonymous users can only access their own cart? There is no [authorization](../sitecore-commerce-security-explained) on data level, meaning that an authenticated user can access all carts. Also, how dow we shield the user from  sensitive data on message level? One way to this is by using an [API Gateway](https://microservices.io/patterns/apigateway.html). An API Gateway allows you to create a new endpoint that proxies/reroutes requests to different service(s). Some of these gateways even allow you to transform the request and response messages.
+
+![Expose through API gateway](/assets/images/expose-commerce-engine/directly.jpg)
+
+Because this project was an experiment, we decided to use the engine directly from JSS to see what impact this has on development speed and complexity of the solution. To be able to select an API gateway we needed to determine our requirements. We did this by selecting a couple of key scenario's and work out how we wanted to expose the relate API's. As an example, the API for retrieving a visitor's cart is displayed below: 
+
+![Cart flow](/assets/images/expose-commerce-engine/cartflow.jpg)
+ 
+Normally the client application authenticates with the commerce engine and can retrieve any cart by its ID. In our case we wanted some mechanism to "anonymously authenticate" (What?) the user. This sounds a bit strange, but it is quite common do to this in headless scenario's. The user invokes a endpoint to get unique token and this token is then used for further API calls to operate in the context of that anonymous user. After authenticating we wanted the cart to be available through `/cart/me` and the user to be determined by passing the token in the headers for that request. The API gateway has to be able to map that URL to the related commerce url `https://commerce:5000/api/Carts('{token}')?$expand=Lines` and transform the token from the headers to the downstream url.
```

Comment:
misschien nog beetje uitwijden over het GUID voor de cart dat in JWT zit? De input voor de `...Carts(token)` is dat GUID ipv het token